### PR TITLE
tests: new script to test example application

### PR DESCRIPTION
**Pull request by tiborsimko**
* BETTER Adds new `test_example_app.sh` script that tests provided
  minimal standalone example application used in the documentation.
  Helps to make sure that the documentation is up to date.

Signed-off-by: Tibor Simko <tibor.simko@cern.ch>

### CODE REVIEW

File: tests/test_example_app.sh [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+echo "[INFO] Testing example app..."
+
+set -o errexit
+set -o nounset
+
+scriptpathname=$(cd "$(dirname $0)" && pwd)
+cd $scriptpathname/../examples
+
+echo "[INFO] Testing database creation..."
+mkdir -p instance
```
Comment (jirikuncar) [inline tests/test_example_app.sh]: we should clean at the end

File: tests/test_example_app.sh [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+rm -f instance/app.db
+flask -a app.py db init
```
Comment (jirikuncar) [inline tests/test_example_app.sh]: are you aware of https://github.com/inveniosoftware/invenio-records/blob/master/examples/app.py#L76-L78 ? 

`flask -a app.py db drop/destroy` should be called at the end

**Comment by jirikuncar** [comment]
Probably we should `trap` at least `SIGINT` and `SIGTERM` call clean-up.

**Comment by jirikuncar** [comment]
Another idea would be to use `python` `subprocess` module for execution.

**Pros** we can run it within pytest
**Cons** TBD

File: tests/test_example_app.sh (hunk 2 above)
Comment (tiborsimko) [inline tests/test_example_app.sh]: > are you aware of https://github.com/inveniosoftware/invenio-records/blob/master/examples/app.py#L76-L78 ? 

Sure; and directly useful e.g. on Travis CI where all three DB backends are tested. 

(If you think of the `rm` statement that is useful for the SQLite backend only, it does not do any harm, but we could indeed take it away...)

> flask -a app.py db drop/destroy should be called at the end

OK

**Comment by tiborsimko** [comment]
> Probably we should trap at least SIGINT and SIGTERM call clean-up.

Indeed; done.

> Another idea would be to use python subprocess module for execution.

One advantage of using shell script for this kind of "integration" testing is that it emulates *precisely* the commands we tell the user to do in the example app documentation. Kind of more direct connection than if we'd execute the statements via Python.

> we can run it within pytest

The test is plugged to `run-tests.sh`, which should be sufficient, since this does more testing than pytest, and developers should be used to running it regularly already.

**Comment by lnielsen** [comment]
Shouldn't these tests be made with selenium instead? I.e test setup/teardown code can be in conftest.py and made with e.g. click.CliRunner.

**Comment by tiborsimko** [comment]
@lnielsen I don't think selenium would be necessary (or advantageous) to test a simple `curl` call. The purpose here is to test the standalone example application that we mention in the docs; nothing more. However, for the demo site "integration" testing in the main invenio repository, selenium would be definitely useful.

**Comment by lnielsen** [comment]
I think we need a general solution which works for all our packages and while in this simple case a shell script might be enough I don't think it will work for other packages. An additional issue is that the test script and documentation needs to kept in sync (perhaps the documentation could extract parts from a script).

For testing example app I think we have two options:
1) With UI (which is hard to test using test client): Selenium (see E2E example in Invenio-Accounts for the pattern)
2) Without much UI: Use pytest with test client.

**Comment by tiborsimko** [comment]
> perhaps the documentation could extract parts from a script instead

This was precisely the approach taken with the Invenio installation scripts. However, there, we had a simple *duo* "sphinx docs <-> installation scripts", so the relation was simple.  Here, we have a *trio* "sphinx docs <-> example app <-> test script".  I prefer to leave the documentation in the example app, because this is what people will read.  So the testing script acts here only to test those (dissociatively documented) instructions. (like unit tests do)

> Without much UI: Use pytest with test client.

Generally yes, however we are testing here the example app's documentation instructions, which are provided in shell script.  Hence the usage of shell scripting to test them as well. Anyway, I agree we can change the language to Python for language consistency in the test suite - however I'd then use `subprocess` and friends so that the shell commands are executed as closely as possible to the documentation.  (To faithfully test the shell commands given there.)

**Comment by tiborsimko** [comment]
I just re-pushed the branch with a new commit that is demonstrating what I had in mind for `subprocess`-driven, yet shell-faithful, testing of documentation instructions.

**Comment by lnielsen** [comment]
+1 for using subprocess

**Comment by tiborsimko** [comment]
OK, squashed to keep the Python version only, and re-pushed.

**Pull request merged**